### PR TITLE
Add Complex Custom User Data Example

### PR DESCRIPTION
**Pull request by takameyer**
## What, How & Why?
We have had questions surrounding best practices in using custom user data, rules and triggers.  This example has been expanded to showcase how one would go about doing this. This requires a more complex setup in the backend and this has been documented in the README.  One should be able to deploy this backend and run this app with their generated AppId without any other updates to the configuration.

Along the way a few things were addressed:
* Type errors in the Auth component.
* Update `realm` and `@realm/react` packages to address bugs regarding client reset and custom user data.
* Replace simple custom user data example.
* Make all data show up in the same database.

## ☑️ ToDos

* [ ] 📝 Changelog entry
* [ ] 📝 `Compatibility` label is updated or copied from previous entry
* [ ] 📝 Update `COMPATIBILITY.md`
* [ ] 🚦 Tests
* [ ] 📦 Updated internal package version in consuming `package.json`s (if updating internal packages)
* [ ] 📱 Check the React Native/other sample apps work if necessary
* [ ] 💥 `Breaking` label has been applied or is not necessary

### CODE REVIEW

File: examples/rn-connection-and-error/frontend/app/providers/StoreProvider.tsx [hunk 1]
```diff
@@ -59,9 +58,12 @@ const StoreContext = createContext<StoreContextType>({
  */
 export function StoreProvider({children}: PropsWithChildren) {
   const realm = useRealm();
-  const store = useQuery(Store)[0];
+  const user = useUser<{}, {storeId: BSON.ObjectId}, {}>();
```
Comment (kraenhansen) [inline examples/rn-connection-and-error/frontend/app/providers/StoreProvider.tsx:61]: We need to improve these generic type signatures 😿

**Review by kraenhansen — APPROVED** [review]

File: examples/rn-connection-and-error/frontend/app/providers/StoreProvider.tsx (hunk 1 above)
Comment (takameyer) [reply examples/rn-connection-and-error/frontend/app/providers/StoreProvider.tsx:61]: We need to improve many things regarding types.  I didn't want to poke that dragon in this PR 🐉

File: examples/rn-connection-and-error/README.md [hunk 2]
```diff
@@ -152,6 +151,13 @@ To modify the [log level and logger](https://www.mongodb.com/docs/realm/sdk/reac
 
 For the App Services logs, you can also choose to [forward the logs to a service](https://www.mongodb.com/docs/atlas/app-services/activity/forward-logs/). To read more about monitoring app activity, please see the [docs](https://www.mongodb.com/docs/atlas/app-services/activity/).
 
+### Tying Custom User Data to Schema Rules
+When a new user is created, they are automatically attached to the first store in the collection (or a new one if it doesn't exist).  The schemas for Store, Kiosk and Products have rules set to only be accessible if the users `storeId` field matches with the appropriate field in each model.  There is a function that can be triggered to switch the store for the current user.  Updating the custom user data and refreshing the session will create a client reset, which will automatically update the UI with the newly selected Store.
```
Comment (kneth) [inline examples/rn-connection-and-error/README.md:155]:
```suggestion
When a new user is created, they are automatically attached to the first store in the collection (or a new store if it doesn't exist).  The schemas for Store, Kiosk and Products have rules set to only be accessible if the users `storeId` field matches with the appropriate field in each model.  There is a function that can be triggered to switch the store for the current user.  Updating the custom user data and refreshing the session will trigger a client reset, which will automatically update the UI with the newly selected Store.
```

File: examples/rn-connection-and-error/README.md [hunk 3]
```diff
@@ -152,6 +151,13 @@ To modify the [log level and logger](https://www.mongodb.com/docs/realm/sdk/reac
...
+
+To perform this in the application, after creating a new user and viewing the store page, one can press `Trigger Store Change`, which will call the `switchStore` function. At this point the UI will not be updated, as a refresh of both the session and the current users custom data must be performed.  To do this, press `Refresh Access Token/User Data`.  The refresh of the session will update the rules in the backend and the refresh of the custom user data will update the UI with the correct store Id.
```
Comment (kneth) [inline examples/rn-connection-and-error/README.md:157]:
```suggestion
To switch store in the application, after creating a new user and viewing the store page, one can press `Trigger Store Change`, which will call the `switchStore` function. At this point the UI will not be updated, as a refresh of both the session and the current users custom data must be performed.  To do this, press `Refresh Access Token/User Data`.  The refresh of the session will update the rules in the backend and the refresh of the custom user data will update the UI with the correct store Id.
```

File: examples/rn-connection-and-error/README.md [hunk 4]
```diff
@@ -152,6 +151,13 @@ To modify the [log level and logger](https://www.mongodb.com/docs/realm/sdk/reac
...
+
+If the refresh session does not happen automatically, one can either press `Refresh Session` or click `Disconnect` followed by `Reconnect`.  If this is done without refreshing the user data, no store will be shown as the rules will be updated to the new store, but the UI will not be updated with the new store Id.
```
Comment (kneth) [inline examples/rn-connection-and-error/README.md:159]:
```suggestion
If the refresh of the session does not happen automatically, one can either press `Refresh Session` or click `Disconnect` followed by `Reconnect`.  If this is done without refreshing the user data, no store will be shown as the rules will be updated to the new store, but the UI will not be updated with the new store Id.
```

File: examples/rn-connection-and-error/README.md [hunk 5]
```diff
@@ -204,19 +210,46 @@ To sync data used in this app you must first:
 
 > If you set up your App Services App [via a CLI](#via-a-cli-recommended), you can **skip this step** as these functions should already be defined for you.
 
-We will add a function for forcing a client reset. The function is solely used for demo purposes and should not be used in production. We will also add a function to be run on user creation that adds fields to the user's [custom user data](https://www.mongodb.com/docs/atlas/app-services/users/custom-metadata/) document.
+We will add functions for the following:
+  * Forcing a client reset.
+    * This function is solely used for demo purposes and should not be used in production.
```
Comment (kneth) [inline examples/rn-connection-and-error/README.md:215]:
```suggestion
    * This function is solely used for demo purposes and should **never** be used in production.
```

File: examples/rn-connection-and-error/README.md [hunk 6]
```diff
@@ -204,19 +210,46 @@ To sync data used in this app you must first:
...
+  * Setting the default `storeId` for a user on creation on the associated [custom user data](https://www.mongodb.com/docs/atlas/app-services/users/custom-metadata/) document.
+    * An [authentication trigger](https://www.mongodb.com/docs/atlas/app-services/triggers/authentication-triggers/) will need to be created and configured to call this function.
+  * Switching the associated Store for the current user.
+  * Creating a new store document and getting all store documents.
+    * These are used to check if the demo stores already exist and create them if not.
```
Comment (kneth) [inline examples/rn-connection-and-error/README.md:220]:
```suggestion
    * These documents are used to check if the demo stores already exist and create them if not.
```

File: examples/rn-connection-and-error/backend/functions/switchStore.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+ * Toggle between two stores.  This is run as the system user, since it will need to create
+ * a second store if one does not exist.  The userId is provided as a string.
+ */
+exports = async function switchStore() {
+  const userId = context.user.id;
+  const customUserDataCollection = context.services.get("mongodb-atlas").db("StoreDemo").collection("Users");
+  const storeCollection = context.services.get("mongodb-atlas").db("StoreDemo").collection("Store");
+
+  try {
+
```
Comment (kneth) [inline examples/rn-connection-and-error/backend/functions/switchStore.js:13]: Can be removed

**Review by kneth — APPROVED** [review]
Minor suggestions for the documentation

File: examples/rn-connection-and-error/README.md (hunk 6 above)
Comment (takameyer) [reply examples/rn-connection-and-error/README.md:220]: I actually I'm referring to the "methods"

File: examples/rn-connection-and-error/backend/functions/switchStore.js (hunk 7 above)
Comment (takameyer) [reply examples/rn-connection-and-error/backend/functions/switchStore.js:13]: What can be removed?  The try catch?

**Review by elle-j — COMMENTED** [review]
Super neat to get these updates in! Great idea to include it in the example 🥳 Rather than providing comments here I [created a complimentary PR](https://github.com/realm/realm-js/pull/6275) to help update certain parts.

File: examples/rn-connection-and-error/backend/data_sources/mongodb-atlas/default_rule.json [hunk 8]
```diff
@@ -1,30 +1 @@
...
-                "read": true
-            },
-            "read": true,
-            "write": false,
-            "insert": false,
-            "delete": false,
-            "search": true
-        }
-    ]
-}
```
[the remark this answers is not in the document]
Comment (takameyer) [reply examples/rn-connection-and-error/backend/data_sources/mongodb-atlas/default_rule.json:30]: Yeah that's a good idea.
I removed the defaults as I had thought they were overriding my custom ones (which the probably weren't)

**Pull request merged**